components/net/image_cache.rs

```diff
@@ -433,33 +487,133 @@ struct ImageCacheStore {
+            }
+            // If we still have images pending keys we need to request a new batch of keys
+            if self.key_cache.images_pending_keys.is_empty() {
+                if let Some(pipeline_id) = self.pipeline_id {
+                    self.compositor_api.generate_image_key_async(pipeline_id);
+                    self.key_cache.cache = KeyCacheState::PendingBatch;
+                } else {
+                    error!("We need image keys for an image_cache without a pipeline");
+                }
+            }
+        } else {
+            unreachable!("A batch was received while we didn't request one")
+        }
+    }
+
+    /// Complete the loading the of the rasterized svg image. This needs the `RasterImage` to
+    /// already have a `WebRenderImageKey`.
+    fn complete_load_svg(
+        &mut self,
+        rasterized_image: RasterImage,
+        pending_image_id: PendingImageId,
+        requested_size: DeviceIntSize,
+    ) {
+        let listeners = {
+            self.rasterized_vector_images
+                .get_mut(&(pending_image_id, requested_size))
+                .map(|task| {
+                    task.result = Some(rasterized_image);
+                    std::mem::take(&mut task.listeners)
+                })
+                .unwrap_or_default()
+        };
+
+        for (pipeline_id, sender) in listeners {
+            let _ = sender.send(ImageCacheResponseMessage::VectorImageRasterizationComplete(
+                RasterizationCompleteResponse {
+                    pipeline_id,
+                    image_id: pending_image_id,
+                    requested_size,
+                },
+            ));
+        }
+    }
+
+    /// The rest of complete load. This requires that images have a valid `WebRenderImageKey`.
+    fn complete_load(&mut self, key: LoadKey, load_result: LoadResult) {
         debug!("Completed decoding for {:?}", load_result);
         let pending_load = match self.pending_loads.remove(&key) {
             Some(load) => load,
             None => return,
         };
 
-        match load_result {
-            LoadResult::LoadedRasterImage(ref mut raster_image) => {
-                set_webrender_image_key(&self.compositor_api, raster_image)
-            },
-            LoadResult::LoadedVectorImage(ref vector_image) => {
-                self.vector_images.insert(key, vector_image.clone());
-            },
-            LoadResult::PlaceholderLoaded(..) | LoadResult::None => {},
-        }
-
         let url = pending_load.final_url.clone();
         let image_response = match load_result {
             LoadResult::LoadedRasterImage(raster_image) => {
-                ImageResponse::Loaded(Image::Raster(Arc::new(raster_image)), url.unwrap())
+                if raster_image.id.is_some() {
+                    ImageResponse::Loaded(Image::Raster(Arc::new(raster_image)), url.unwrap())
+                } else {
+                    ImageResponse::None
+                }
```

Comment:
The conditional here changes the logic of the cache(introducing the branch returning `                    ImageResponse::None` which is absent on `main`), but it is has no effect because here the `id`will always be `Some`, so we should remove the condition(I would say even `assert!(is_some())`).